File: kotlin-semver/karakum/plugins/convertVersionPartTypes.js
import ts from "typescript";

const versionParts = new Set(["major", "minor", "patch"])

export default function (node) {
    if (
        node.kind === ts.SyntaxKind.NumberKeyword

        && node.parent
        && ts.isPropertyDeclaration(node.parent)

        && node.parent.parent
        && ts.isClassDeclaration(node.parent.parent)

        && node.parent.parent.name?.text === "SemVer"

        && ts.isIdentifier(node.parent.name)
        && versionParts.has(node.parent.name.text)
    ) {
        return "Int"
    }

    if (
        node.kind === ts.SyntaxKind.NumberKeyword

        && node.parent
        && ts.isFunctionDeclaration(node.parent)
        && node.parent.type === node

        && ts.isIdentifier(node.parent.name)
        && versionParts.has(node.parent.name.text)
    ) {
        return "Int"
    }

    return null
}

File: kotlin-semver/karakum/src/jsMain/kotlin/semver/karakum/plugins/convertVersionPartTypes.kt
package semver.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.SyntaxKind
import typescript.isClassDeclaration
import typescript.isFunctionDeclaration
import typescript.isIdentifier
import typescript.isPropertyDeclaration

private val versionParts = setOf("major", "minor", "patch")

val convertVersionPartTypes = createPlugin { node, _, _ ->
    nullable {
        ensure(node.kind == SyntaxKind.NumberKeyword)

        val property = ensureNotNull(node.getParentOrNull())
        ensure(isPropertyDeclaration(property))

        val propertyName = property.name
        ensure(isIdentifier(propertyName))
        ensure(propertyName.text in versionParts)

        val classNode = ensureNotNull(property.getParentOrNull())
        ensure(isClassDeclaration(classNode))
        ensure(classNode.name?.text == "SemVer")

        "Int"
    } ?: nullable {
        ensure(node.kind == SyntaxKind.NumberKeyword)

        val function = ensureNotNull(node.getParentOrNull())
        ensure(isFunctionDeclaration(function))
        ensure(function.name?.text in versionParts)
        ensure(function.type === node)

        "Int"
    }
}
